### PR TITLE
Restructure processors to work with mixed Ory+Istio handlers

### PR DESCRIPTION
**Pull request by videlov**
**Description**

Changes proposed in this pull request:

- Restructure processors to cover mixed cases for Istio and Ory when FT is on `istio`

**Related issue(s)**

https://github.com/kyma-project/api-gateway/issues/96
https://github.com/kyma-project/api-gateway/issues/75

### CODE REVIEW

File: internal/processing/istio/access_rule_processor.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	defaultDomainName string
+}
+
+// Create returns a map of rules using the configuration of the APIRule. The key of the map is a unique combination of
+// the match URL and methods of the rule.
+func (r accessRuleCreator) Create(api *gatewayv1beta1.APIRule) map[string]*rulev1alpha1.Rule {
+	pathDuplicates := processors.HasPathDuplicates(api.Spec.Rules)
+	accessRules := make(map[string]*rulev1alpha1.Rule)
+	for _, rule := range api.Spec.Rules {
+		filterFunc := func(auth *gatewayv1beta1.Authenticator) bool { return auth.Handler.Name != "jwt" }
```
[the remark this answers is not in the document]
Comment (videlov) [reply internal/processing/istio/access_rule_processor.go:42]: We need the list of filtered access strategies so we create access rules only for those. Will adapt it to a helper function ofc.

File: internal/processing/istio/access_rule_processor.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		filteredAS := processing.FilterGeneric(rule.AccessStrategies, filterFunc)
+		if len(filteredAS) > 0 && processing.IsSecured(rule) {
```
[the remark this answers is not in the document]
Comment (videlov) [reply internal/processing/istio/access_rule_processor.go:44]: Yes, we need it. We have this mixed scenario when we have Istio JWT and Ory handler on different paths.

File: internal/processing/istio/access_rule_processor_test.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	. "github.com/onsi/gomega"
+	rulev1alpha1 "github.com/ory/oathkeeper-maester/api/v1alpha1"
+	networkingv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var _ = Describe("Access Rule Processor", func() {
+	When("handler is allow", func() {
```
[the remark this answers is not in the document]
Comment (videlov) [reply internal/processing/istio/access_rule_processor_test.go:21]: Will include oauth2 test for istio access rule processor.

File: internal/processing/istio/reconciliation_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+			// when
+			var createdObjects []client.Object
+			reconciliation := istio.NewIstioReconciliation(GetTestConfig())
+			for _, processor := range reconciliation.GetProcessors() {
+				results, err := processor.EvaluateReconciliation(context.TODO(), faceClient, apiRule)
+				Expect(err).To(BeNil())
+
+				for _, result := range results {
+					_, vsOk := result.Obj.(*networkingv1beta1.VirtualService)
+					_, arOk := result.Obj.(*rulev1alpha1.Rule)
```
[the remark this answers is not in the document]
Comment (videlov) [reply internal/processing/istio/reconciliation_test.go:58]: Absolutely this makes sense. Will add it.

File: internal/processing/istio/reconciliation_test.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+			Expect(vsCreated && raCreated && apCreated == 2).To(BeTrue())
+		})
+
+		It("with Ory oauth2 should provide Istio VS, AP, RA and Ory rule", func() {
+			// given
+			oauthConfigJSON := fmt.Sprintf(`{"required_scope": [%s]}`, ToCSVList(ApiScopes))
+			oauth := []*gatewayv1beta1.Authenticator{
+				{
+					Handler: &gatewayv1beta1.Handler{
+						Name: "noop",
```
Comment (triffer) [inline internal/processing/istio/reconciliation_test.go:99]: The test description is about oauth2, but the handler user is noop. Also the oauthConfigJSON is used.

File: internal/processing/istio/authorization_policy_processor_test.go [hunk 6]
```diff
@@ -194,4 +196,172 @@ var _ = Describe("Authorization Policy Processor", func() {
 		Expect(ap.OwnerReferences[0].Name).To(Equal(ApiName))
 		Expect(ap.OwnerReferences[0].UID).To(Equal(ApiUID))
 	})
+
+	When("single handler only", func() {
+		It("should not create AP for allow", func() {
```
Comment (triffer) [inline internal/processing/istio/authorization_policy_processor_test.go:201]: We could add a separate test to verify that APs created for jwt handlers contain a Spec.Rules.From.

File: internal/builders/istio.go [hunk 7]
```diff
@@ -113,6 +113,13 @@ func (r *Rule) RuleFrom(val *RuleFrom) *Rule {
 	return r
 }
 
+func (r *Rule) RuleFromConditional(val *RuleFrom, condition bool) *Rule {
```
Comment (triffer) [inline internal/builders/istio.go:116]: I think having a conditional in the builder is not a good thing. I this case we make it generic, but that makes it actually harder to follow in what case this is relevant. I think as we have a builder we can move the conditional closer to the actual logic and only invoke the RuleFrom conditional, e.g. 

in generateAuthorizationPolicySpec
```
	ruleBuilder := builders.RuleBuilder().RuleTo(builders.RuleToBuilder().
		Operation(builders.OperationBuilder().Methods(rule.Methods).Path(rule.Path)))

	if processing.IsJwtSecured(rule) {
		ruleBuilder.RuleFrom(builders.RuleFromBuilder().Source())
	}

```

File: internal/processing/istio/reconciliation_test.go (hunk 5 above)
Comment (videlov) [reply internal/processing/istio/reconciliation_test.go:99]: Yes, this is copy-paste typo.

File: internal/builders/istio.go (hunk 7 above)
Comment (videlov) [reply internal/builders/istio.go:116]: Yes, agree. I was also unsure about having a conditional param within builders.

**Review by triffer — APPROVED** [review]

**Pull request merged**